deepinv/datasets/datagenerator.py

```diff
@@ -58,69 +95,168 @@ def __init__(
         dtype: torch.dtype = torch.float,
         complex_dtype: torch.dtype = torch.cfloat,
     ):
+        import h5py
+
         super().__init__()
-        self.data_info = []
-        self.data_cache = {}
-        self.unsupervised = False
+
+        f = h5py.File(path, "r")
+
+        self._register_members(
+            f,
+            train=train,
+            split=split,
+            load_physics_generator_params=load_physics_generator_params,
+        )
+
+        self.hd5 = f
         self.transform = transform
-        self.load_physics_generator_params = load_physics_generator_params
         self.cast = lambda x: x.type(complex_dtype if x.is_complex() else dtype)
 
-        if isinstance(h5py, ImportError):
-            raise h5py
-
-        self.hd5 = h5py.File(path, "r")
-        suffix = ("_train" if train else "_test") if split is None else f"_{split}"
+        self._register_deprecated_attributes()
 
-        if "stacked" in self.hd5.attrs.keys():
-            self.stacked = self.hd5.attrs["stacked"]
-            self.y = [self.hd5[f"y{i}{suffix}"] for i in range(self.stacked)]
+    def _register_members(
+        self,
+        f: Any,
+        *,
+        train: bool,
+        split: str | None,
+        load_physics_generator_params: bool,
+    ) -> None:
+        # Process ground truths
+        x = None
+
+        # Process measurements
+        attrs: dict = f.attrs
+        marked_stacked = "stacked" in attrs
+        stacked = attrs.get("stacked", 0)
+        if marked_stacked:
+            y = [None] * stacked
         else:
-            self.stacked = 0
-            self.y = self.hd5[f"y{suffix}"]
+            y = None
 
-        if train or split == "train":
-            if "x_train" in self.hd5:
-                self.x = self.hd5[f"x{suffix}"]
-            else:
-                self.unsupervised = True
+        # Process forward operator parameters
+        if load_physics_generator_params:
+            params = {}
         else:
-            self.x = self.hd5[f"x{suffix}"]
+            params = None
+
+        # Register members in the HDF5 file as ground truths, parts of stacked
+        # measurements or regular measurements and forward operator parameters
+        split_name = split if split is not None else ("train" if train else "test")
```

Comment:
should we add a warning when the arg is overridden to prevent jump-scares in future?